packet/swcksum.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2017 Intel Corporation.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package packet
+
+import (
+	. "github.com/intel-go/yanff/common"
+	"unsafe"
+)
+
+// Calculates checksum of memory for a given pointer. Length and
+// offset are in bytes. Offset is signed, so negative offset is
+// possible. Checksum is calculated in uint16 words. Returned is
+// checksum with carry, so carry should be added and value negated for
+// use as network checksum.
+func calculateDataChecksum(ptr unsafe.Pointer, length, offset int) uint32 {
+	var sum uint32 = 0
+	uptr := uintptr(ptr) + uintptr(offset)
+
+	slice := (*[1 << 30]uint16)(unsafe.Pointer(uptr))[0 : length/2]
+	for i := range slice {
+		sum += uint32(SwapBytesUint16(slice[i]))
+	}
+
+	if length&1 != 0 {
+		sum += uint32(*(*byte)(unsafe.Pointer(uptr + uintptr(length-1)))) << 8
+	}
+
+	return sum
+}
+
+func CalculatePseudoHdrIPv4TCPCksum(hdr *IPv4Hdr) uint16 {
```

Comment:
We need to add documentation for public functions in this file, those that start with capital letter.